### ISSUE TITLE
check_ping to local FQDN fails with CentOS 7.4

### ISSUE THREAD

**Issue by peteeckel** [nagios-plugins/nagios-plugins#326]
After upgrading from CentOS 7.3 to CentOS 7.4, there is a change in behaviour of check_ping which causes problems in some installations. 

When check_ping is executed against the FQDN of the local machine running CentOS 7.3 and earlier, check_ping used to refer to one of the link-local addresses of the machine and succeed: 

    [root@centos73 ~]# /usr/lib64/nagios/plugins/check_ping -H centos73.example.com -w 3000,80% -c 5000,100%
    PING OK - Packet loss = 0%, RTA = 0.05 ms|rta=0.050000ms;3000.000000;5000.000000;0.000000 pl=0%;80;100;0

    [root@centos74 ~]# /usr/lib64/nagios/plugins/check_ping -H centos74.example.com -w 3000,80% -c 5000,100%
    /usr/bin/ping6 -n -U -W 30 -c 5 centos74.example.com
    CRITICAL - Could not interpret output from ping command

Neither of the machines has an IPv6 address configured for the FQDN. The root cause is likely that check_ping uses /usr/bin/ping6 for some reason anyway, and the behaviour of ping6 has changed in CentOS 7.4:

    [root@centos73 ~]# ping6 centos73.example.com
    PING centos73.example.com(centos73.example.com (fe80::a00:27ff:fef1:95c4)) 56 data bytes
    64 bytes from centos73.example.com (fe80::a00:27ff:fef1:95c4%enp0s8): icmp_seq=1 ttl=64 time=0.054 ms
    64 bytes from centos73.example.com (fe80::a00:27ff:fef1:95c4%enp0s8): icmp_seq=2 ttl=64 time=0.040 ms
    64 bytes from centos73.example.com (fe80::a00:27ff:fef1:95c4%enp0s8): icmp_seq=3 ttl=64 time=0.038 ms
    [...]

vs. 

    [root@centos74 ~]# ping6 centos74.example.com
    connect: Invalid argument

While the root cause is likely to be a problem in RHEL 7.4 and should be handled by RedHat (unfortunately I don't have an account with the entitlement to open issues with RHEL), I keep asking myself why check_ping resorts to ping6 anyway when there is no IPv6 address configured for the FQDN. I did a quick strace and found that check_ping even tests for the value of 

    /proc/sys/net/ipv6/conf/all/disable_ipv6

before running the check, but then uses ping6 anyway even if IPv6 is disabled.

**Comment by box293** [comment]
I can't reproduce this, I suspect there is an issue with the symbolic links for ping6 -> ping.

Here are my ping examples:
```
[root@centos10 ~]# cat /etc/centos-release
CentOS Linux release 7.3.1611 (Core)
[root@centos10 ~]# host centos10.box293.local
centos10.box293.local has address 10.25.13.15
[root@centos10 ~]# ping6 centos10.box293.local
PING centos10.box293.local(centos10.box293.local (2001:44b8:3132:25:10:25:254:931a)) 56 data bytes
64 bytes from centos10.box293.local (2001:44b8:3132:25:10:25:254:931a): icmp_seq=1 ttl=64 time=0.047 ms
```

Then I upgrade and reboot
```
[root@centos10 ~]# cat /etc/centos-release
CentOS Linux release 7.4.1708 (Core) 
[root@centos10 ~]# ping6 centos10.box293.local
PING centos10.box293.local(centos10.box293.local (2001:44b8:3132:25:10:25:254:931a)) 56 data bytes
64 bytes from centos10.box293.local (2001:44b8:3132:25:10:25:254:931a): icmp_seq=1 ttl=64 time=0.042 ms
64 bytes from centos10.box293.local (2001:44b8:3132:25:10:25:254:931a): icmp_seq=2 ttl=64 time=0.044 ms
^C
--- centos10.box293.local ping statistics ---
2 packets transmitted, 2 received, 0% packet loss, time 1001ms
rtt min/avg/max/mdev = 0.042/0.043/0.044/0.001 ms
[root@centos10 ~]# host centos10.box293.local
centos10.box293.local has address 10.25.13.15
```

What is the output of the following commands:
```
which -a ping
which -a ping | xargs -r ls -la

which -a ping6
which -a ping6 | xargs -r ls -la

grep '^[^#\]' /etc/nsswitch.conf

host centos73.example.com
```
The `host` command requires bind-utils to be installed.

**Comment by peteeckel** [comment]
First of all, your test case doesn't fully reflect mine. 

    [root@centos10 ~]# ping6 centos10.box293.local
    PING centos10.box293.local(centos10.box293.local (2001:44b8:3132:25:10:25:254:931a)) 56 data bytes
    64 bytes from centos10.box293.local (2001:44b8:3132:25:10:25:254:931a): icmp_seq=1 ttl=64 time=0.047 ms

Your host obviously has a global IPv6 address, probably assigned via SLAAC. Here's a Vagrantfile that I used to reproduce the issue: 

    Vagrant.configure("2") do |config|
        config.vm.define "centos73" do |centos73|
            centos73.vm.box = "bento/centos-7.3"
            centos73.vm.hostname = "centos73.example.com"

            centos73.vm.provision "shell", inline: "yum -y --noplugins --nogpgcheck install epel-release"
            centos73.vm.provision "shell", inline: "yum -y --noplugins --nogpgcheck install nagios-plugins-all"
        end
    
        config.vm.define "centos74" do |centos74|
            centos74.vm.box = "bento/centos-7.4"
            centos74.vm.hostname = "centos74.example.com"

            centos74.vm.provision "shell", inline: "yum -y --noplugins --nogpgcheck install epel-release"
            centos74.vm.provision "shell", inline: "yum -y --noplugins --nogpgcheck install nagios-plugins-all"
        end
    end

With this setup I do **not** get a global IPv6 address on either machine, reflecting the state of the production box the problem initially occurred on: 

    [vagrant@centos73 ~]$ ip address
    1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN qlen 1
        link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
        inet 127.0.0.1/8 scope host lo
           valid_lft forever preferred_lft forever
        inet6 ::1/128 scope host 
           valid_lft forever preferred_lft forever
    2: enp0s3: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP qlen 1000
        link/ether 08:00:27:d0:c3:19 brd ff:ff:ff:ff:ff:ff
        inet 10.0.2.15/24 brd 10.0.2.255 scope global dynamic enp0s3
           valid_lft 85579sec preferred_lft 85579sec
        inet6 fe80::b2c9:f476:ca94:907/64 scope link 
           valid_lft forever preferred_lft forever

    [vagrant@centos74 ~]$ ip address 
    1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN qlen 1
        link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
        inet 127.0.0.1/8 scope host lo
           valid_lft forever preferred_lft forever
        inet6 ::1/128 scope host 
           valid_lft forever preferred_lft forever
    2: enp0s3: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP qlen 1000
        link/ether 08:00:27:96:27:00 brd ff:ff:ff:ff:ff:ff
        inet 10.0.2.15/24 brd 10.0.2.255 scope global dynamic enp0s3
           valid_lft 85584sec preferred_lft 85584sec
        inet6 fe80::1029:8cd6:517:62bf/64 scope link 
           valid_lft forever preferred_lft forever

The problem can be reproduced now: 

    [vagrant@centos73 ~]$ ping6 -c 1 centos73.example.com
    PING centos73.example.com(centos73.example.com (fe80::b2c9:f476:ca94:907)) 56 data bytes
    64 bytes from centos73.example.com (fe80::b2c9:f476:ca94:907%enp0s3): icmp_seq=1 ttl=64 time=0.135 ms
    
    --- centos73.example.com ping statistics ---
    1 packets transmitted, 1 received, 0% packet loss, time 0ms
    rtt min/avg/max/mdev = 0.135/0.135/0.135/0.000 ms

    [vagrant@centos74 ~]$ ping6 -c 1 centos74.example.com
    connect: Invalid argument

Which shows that obviously the behaviour of ping6 has changed for this setup, which is probably a RHEL issue. Running the tests you requested gives the following output on CentOS 7.3: 

    [vagrant@centos73 ~]$ which -a ping
    /usr/bin/ping
    
    [vagrant@centos73 ~]$ which -a ping | xargs -r ls -la
    -rwxr-xr-x. 1 root root 62088 Nov  7  2016 /usr/bin/ping
    
    [vagrant@centos73 ~]$ which -a ping6
    /usr/bin/ping6
    /usr/sbin/ping6
    
    [vagrant@centos73 ~]$ which -a ping6 | xargs -r ls -la
    lrwxrwxrwx. 1 root root  4 Jun 26 19:26 /usr/bin/ping6 -> ping
    lrwxrwxrwx. 1 root root 11 Jun 26 19:26 /usr/sbin/ping6 -> ../bin/ping
    
    [vagrant@centos73 ~]$ grep '^[^#\]' /etc/nsswitch.conf
    passwd:     files sss
    shadow:     files sss
    group:      files sss
    hosts:      files dns myhostname
    bootparams: nisplus [NOTFOUND=return] files
    ethers:     files
    netmasks:   files
    networks:   files
    protocols:  files
    rpc:        files
    services:   files sss
    netgroup:   files sss
    publickey:  nisplus
    automount:  files
    aliases:    files nisplus
    
    [vagrant@centos73 ~]$ host centos73.example.com
    Host centos73.example.com not found: 3(NXDOMAIN)
    
And on CentOS 7.4:    

    [vagrant@centos74 ~]$ which -a ping
    /usr/bin/ping
    
    [vagrant@centos74 ~]$ which -a ping | xargs -r ls -la
    -rwxr-xr-x. 1 root root 66176 Aug  4 08:01 /usr/bin/ping
    
    [vagrant@centos74 ~]$ which -a ping6
    /usr/bin/ping6
    /usr/sbin/ping6
    
    [vagrant@centos74 ~]$ which -a ping6 | xargs -r ls -la
    lrwxrwxrwx. 1 root root  4 Sep 15 20:30 /usr/bin/ping6 -> ping
    lrwxrwxrwx. 1 root root 11 Sep 15 20:30 /usr/sbin/ping6 -> ../bin/ping
    
    [vagrant@centos74 ~]$ grep '^[^#\]' /etc/nsswitch.conf
    passwd:     files sss
    shadow:     files sss
    group:      files sss
    hosts:      files dns myhostname
    bootparams: nisplus [NOTFOUND=return] files
    ethers:     files
    netmasks:   files
    networks:   files
    protocols:  files
    rpc:        files
    services:   files sss
    netgroup:   nisplus sss
    publickey:  nisplus
    automount:  files nisplus sss
    aliases:    files nisplus
    
    [vagrant@centos74 ~]$ host centos74.example.com
    Host centos74.example.com not found: 3(NXDOMAIN)

**Comment by peteeckel** [comment]
Just to rule out possible issues with the CentOS 7.4 Vagrant box, I just updated the 7.3 box to 7.4 and rebooted. Now the issue can be reproduced on the former 7.3 machine as well: 

    [vagrant@centos73 ~]$ cat /etc/redhat-release 
    CentOS Linux release 7.4.1708 (Core) 

    [vagrant@centos73 ~]$ ping6 -c 1 centos73.example.com
    connect: Invalid argument

**Comment by box293** [comment]
Thanks for the extra testing, it should help the DEVs pinpoint the issue.

**Comment by peteeckel** [comment]
Unfortunately, it did not. 

The issue is still around with CentOS 7.5 and the latest nagios-plugins (2.2.1-9git5c7eb5b9.el7) version. I don't think https://github.com/Icinga/icinga2/issues/6093 is relevant, since the problem lies with the OS and the way check_ping uses it - worked in CentOS/RHEL <= 7.3, broken in 7.4 and higher.

**Comment by tokiwinter** [comment]
Same issue here, workaround suggested by @rojobull (simply adding `-4` to my `check-host-alive` command) fixes things for me.

```
# cat /etc/os-release
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"
# rpm -q nagios nagios-plugins-ping
nagios-4.3.4-5.el7.x86_64
nagios-plugins-ping-2.2.1-9git5c7eb5b9.el7.x86_64
# nagios --version

Nagios Core 4.3.4
Copyright (c) 2009-present Nagios Core Development Team and Community Contributors
Copyright (c) 1999-2009 Ethan Galstad
Last Modified: 2017-08-24
License: GPL

Website: https://www.nagios.org
This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License version 2 as
published by the Free Software Foundation.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
# /usr/lib64/nagios/plugins/check_ping --version
check_ping v2.2.1.git (nagios-plugins 2.2.1)
```

**Comment by box293** [comment]
> Just to rule out possible issues with the CentOS 7.4 Vagrant box, I just updated the 7.3 box to 7.4 and rebooted. Now the issue can be reproduced on the former 7.3 machine as well:
> 
> [vagrant@centos73 ~]$ cat /etc/redhat-release 
> CentOS Linux release 7.4.1708 (Core) 
> 
> [vagrant@centos73 ~]$ ping6 -c 1 centos73.example.com
> connect: Invalid argument


I was able to reproduce the `connect: Invalid argument` error:

```
[root@centos19 ~]# ip addr
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN qlen 1
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: ens32: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP qlen 1000
    link/ether 00:50:56:ab:41:93 brd ff:ff:ff:ff:ff:ff
    inet 10.25.13.37/16 brd 10.25.255.255 scope global ens32
       valid_lft forever preferred_lft forever
    inet6 fe80::7407:f228:1a4c:1e0e/64 scope link 
       valid_lft forever preferred_lft forever

[root@centos19 ~]# ping6 -c 1 fe80::7407:f228:1a4c:1e0e
connect: Invalid argument

[root@centos19 ~]# ping6 -c 1 fe80::7407:f228:1a4c:1e0e%ens32
PING fe80::7407:f228:1a4c:1e0e%ens32(fe80::7407:f228:1a4c:1e0e%ens32) 56 data bytes
64 bytes from fe80::7407:f228:1a4c:1e0e%ens32: icmp_seq=1 ttl=64 time=0.040 ms

--- fe80::7407:f228:1a4c:1e0e%ens32 ping statistics ---
1 packets transmitted, 1 received, 0% packet loss, time 0ms
rtt min/avg/max/mdev = 0.040/0.040/0.040/0.000 ms
```

Posting this here so I don't forget it.

**Comment by box293** [comment]
So I think I understand what the problem is. When the ping command is using an IPv6 link local address, it doesn't define the interface that the address is for like `ens32`, this needs to be defined by the end user. Here's some examples:

```
[root@centos19 ~]# head -n1 /etc/hosts
fe80::1d2f:f36c:7785:d500 localhostx

[root@centos19 ~]# ping6 -c 1 localhostx
connect: Invalid argument

[root@centos19 ~]# ping6 -c 1 localhostx -I ens32
PING localhostx(localhostx (fe80::1d2f:f36c:7785:d500)) from fe80::7407:f228:1a4c:1e0e%ens32 ens32: 56 data bytes
64 bytes from localhostx (fe80::1d2f:f36c:7785:d500%ens32): icmp_seq=1 ttl=64 time=0.148 ms

--- localhostx ping statistics ---
1 packets transmitted, 1 received, 0% packet loss, time 0ms
rtt min/avg/max/mdev = 0.148/0.148/0.148/0.000 ms
```

`check_ping` is not doing validation on the resolved address except to ensure it's a valid address using `getaddrinfo`.

What is the solution for this?

* Add an interface  `-I ` argument to be able to define which interface is used?
  * Requires command changes
* Add some address resolution and interface checking so that the plugin automatically determines the correct -I argument to supply to the ping command?
  * Adds a lot more logic to the plugin including address resolution which may be overkill

What are your thoughts on this?

**Comment by mtnielsen** [comment]
Pragmatic me says defaulting to IPv4 in check_ping is the right thing to do, as this ought to work in any given scenario, but the idealistic me says IPv6 is better so default to that and to disregard the backwards bumpkins who don't have IPv6 yet.

I think pragmatic me is right in this case though as a Nagios plugin is not the right place for a 4v6 battle. Defaulting to the option that works in most scenarios is the better option.
